crates/env/src/engine/on_chain/impls.rs

```diff
@@ -272,17 +272,26 @@ impl EnvBackend for EnvInstance {
         <H as CryptoHash>::hash(enc_input, output)
     }
 
-    #[cfg(feature = "ink-unstable-chain-extensions")]
-    fn call_chain_extension<I, O>(&mut self, func_id: u32, input: &I) -> Result<O>
+    fn call_chain_extension<I, T, E, ErrorCode, F, D>(
+        &mut self,
+        func_id: u32,
+        input: &I,
+        status_to_result: F,
+        decode_to_result: D,
+    ) -> ::core::result::Result<T, E>
     where
         I: scale::Encode,
-        O: scale::Decode,
+        T: scale::Decode,
+        E: From<ErrorCode>,
+        F: FnOnce(u32) -> ::core::result::Result<(), ErrorCode>,
+        D: FnOnce(&[u8]) -> ::core::result::Result<T, E>,
     {
         let mut scope = self.scoped_buffer();
         let enc_input = scope.take_encoded(input);
         let output = &mut scope.take_rest();
-        ext::call_chain_extension(func_id, enc_input, output)?;
-        scale::Decode::decode(&mut &output[..]).map_err(Into::into)
+        status_to_result(ext::call_chain_extension(func_id, enc_input, output))?;
+        let decoded = decode_to_result(&mut &output[..])?;
```

Comment:
Noting that this pattern of `status_to_result()?` then `decode_to_result()?` is required in both on and off chain impls to satisfy the API, as referred to in previous comment. Not sure it's worth extracting but worth noting at least on the trait method comment.

Reply:
Good point, I added a `# Dev. Note` to the `Backend::call_chain_extension` method to state the implementation requirements.